Implemented output of reservoir statistics

francoishamon opened the pull request:
This PR adds some output for quantities that we have been asked to report for comparison and analysis purposes. These quantities are computed for each `ElementRegions` and can be accessed using `TimeHistory`.

The PR is ready for review, and will require a rebaseline. Rebaseline done in https://github.com/GEOSX/integratedTests/pull/243

Resolves #1968

TotoGaz commented on src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseBase.cpp:
```diff
@@ -859,8 +868,161 @@ void CompositionalMultiphaseBase::initializePostInitialConditionsPreSubGroups()
+  m_averagePressure /= totalUncompactedPoreVolume;
+  m_averageTemperature /= totalUncompactedPoreVolume;
+
+  if( outputToTerminal )
```
Are you computing all the above even if you do not display?
Are you using the information somewhere else?
If  possible, I would make this computation iff `outputToTerminal` is `true`.

francoishamon replied:
Same here, the information will be always computed when `computeStatistics` is `true`, but only accessible through `TimeHistory`

TotoGaz commented on src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseBase.cpp:
```diff
@@ -106,6 +100,18 @@ CompositionalMultiphaseBase::CompositionalMultiphaseBase( const string & name,
 
+  // Statistics
+
+  this->registerWrapper( viewKeyStruct::averagePressureString(), &m_averagePressure );
+  this->registerWrapper( viewKeyStruct::minimumPressureString(), &m_minimumPressure );
+  this->registerWrapper( viewKeyStruct::maximumPressureString(), &m_maximumPressure );
+  this->registerWrapper( viewKeyStruct::averageTemperatureString(), &m_averageTemperature );
+  this->registerWrapper( viewKeyStruct::minimumTemperatureString(), &m_minimumTemperature );
+  this->registerWrapper( viewKeyStruct::maximumTemperatureString(), &m_maximumTemperature );
+  this->registerWrapper( viewKeyStruct::totalPoreVolumeString(), &m_totalPoreVolume );
+  this->registerWrapper( viewKeyStruct::phasePoreVolumeString(), &m_phasePoreVolume ).
+    setSizedFromParent( 0 );
```
You are attaching/registering/giving a name to 8 member to your solver just to display some convergence information. That sounds heavy.

I would preferably build a `struct` with the proper fields
```c++
struct{
  real64 m_averagePressure;
  real64 m_minimumPressure;
  real64 m_maximumPressure;
  real64 m_averageTemperature;
  real64 m_minimumTemperature;
  real64 m_maximumTemperature;
  real64 m_totalPoreVolume;
  array1d< real64 > m_phasePoreVolume;
}
```
and get your `computeReservoirStatistics` build and return it. Then you could display it in your `computeStatistics`.

francoishamon replied:
To simplify the code, I removed the `averageTemperature`, `minTemperature`, and `maxTemperature` so there are not fewer quantities to register. Also, I removed all the member variables `m_X` used for statistics (I just register data on the mesh, without using member variables).

TotoGaz commented on src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseBase.cpp:
```diff
@@ -859,8 +868,161 @@ void CompositionalMultiphaseBase::initializePostInitialConditionsPreSubGroups()
+  }
+
+  // output the average pressure, average temperature, pore volumes
+  computeReservoirStatistics( mesh, outputStatisticsToScreen );
```
Move this `computeReservoirStatistics( mesh, outputStatisticsToScreen );` into the `if ( outputStatisticsToScreen...` scope (I think).

francoishamon replied:
I am no longer going to output anything to terminal (average pressure, nonlinear iterations, etc), so I got rid of `outputStatisticsToScreen`.

TotoGaz commented on src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseBase.cpp:
```diff
@@ -188,7 +194,7 @@ void CompositionalMultiphaseBase::registerDataOnMesh( Group & meshBodies )
         reference().resizeDimension< 1, 2 >( m_numPhases, m_numComponents );
 
-      if( m_computeCFLNumbers )
+      if( m_computeStatistics )
```
Does it make sense that `phaseOutflux`, `componentOutflux`, `phaseCFLNumber`, `componentCFLNumber` are conditionally register on `if( m_computeStatistics )`?
I could spot no usage of them related to the statistics display.

francoishamon replied:
Yes, I think it makes sense. Now, the CFL numbers are computed only when `computeStatistics` is true, in which case we have to register these quantities (`phaseOutflux`, `componentOutflux`, etc). If `computeStatistics` is `false`, the CFL numbers are not computed and these quantities are not needed.

TotoGaz commented on src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseBase.cpp:
```diff
@@ -859,8 +868,161 @@ void CompositionalMultiphaseBase::initializePostInitialConditionsPreSubGroups()
+
+  } );
+
+  m_averagePressure = MpiWrapper::sum( m_averagePressure );
+  m_minimumPressure = MpiWrapper::min( m_minimumPressure );
+  m_maximumPressure = MpiWrapper::max( m_maximumPressure );
+
+  m_averageTemperature = MpiWrapper::sum( m_averageTemperature );
+  m_minimumTemperature = MpiWrapper::min( m_minimumTemperature );
+  m_maximumTemperature = MpiWrapper::max( m_maximumTemperature );
+
+  m_totalPoreVolume = 0.0;
+  totalUncompactedPoreVolume = MpiWrapper::sum( totalUncompactedPoreVolume );
+  for( integer ip = 0; ip < m_numPhases; ++ip )
+  {
+    m_phasePoreVolume[ip] = MpiWrapper::sum( m_phasePoreVolume[ip] );
+    m_totalPoreVolume += m_phasePoreVolume[ip];
+  }
+
+  m_averagePressure /= totalUncompactedPoreVolume;
+  m_averageTemperature /= totalUncompactedPoreVolume;
```
Could you regroup `min` and `max` on top, separated form the averages computations (which is more complicated)?

More important, are you sure that you are properly computing the average?
How are you considering, _e.g._ the weighted sum for each domain?
This is not so clear to me, but I may be wrong.

francoishamon replied:
I separated the `min` and `max` from the averages as you suggested.

Yes, I think I am properly computing the averages. The average is computed to reproduce the `FPRP` keyword in our references. Specifically, in the average, the pressure is weighted by the uncompacted pore volume (`referencePorosity * volume`), and in my tests, the GEOSX average pressure matches well `FPRP`.

TotoGaz reviewed (commented):
First review.
Definitely conflicting with https://github.com/GEOSX/GEOSX/pull/1677.
It would be better to call the `displayStatistics` directly from the `SolverBase` loop on top (like https://github.com/GEOSX/GEOSX/pull/1677 does IIRC).

(and maybe rename to `dislpayStatistics` also, but we should discuss about it.)

francoishamon commented:
> First review. Definitely conflicting with #1677. It would be better to call the `displayStatistics` directly from the `SolverBase` loop on top (like #1677 does IIRC).
> 
> (and maybe rename to `dislpayStatistics` also, but we should discuss about it.)

@TotoGaz I changed my mind for this PR. Initially I was planning to modify `SolverBase` and have a mechanism to output the nonlinear iterations to the terminal, which was implemented in the first version of the PR. Now, to finish this quickier, I am going to downsize the PR to just compute the average quantities in `CompositionalMultiphaseBase`, because this is the most urgent request. I no longer plan to modify `SolverBase`. I am going to start implementing your suggestions now

francoishamon commented on src/coreComponents/physicsSolvers/SolverBase.cpp:
```diff
@@ -75,6 +75,16 @@ SolverBase::SolverBase( string const & name,
+  registerWrapper( viewKeyStruct::statisticsOutputFrequencyString(), &m_statisticsOutputFrequency ).
+    setApplyDefaultValue( 10 ).
+    setInputFlag( InputFlags::OPTIONAL ).
+    setDescription( "Flag indicating the frequency of the output of statistics to screen" );
```
I got rid of this output frequency (which was used to output to file), the output frequency will now be decided by `TimeHistory`.

francoishamon commented on src/coreComponents/physicsSolvers/SolverBase.hpp:
```diff
@@ -494,6 +494,24 @@ class SolverBase : public ExecutableGroup
+                     real64 const & dt,
+                     integer cycleNumber,
+                     DomainPartition & domain,
+                     bool outputStatisticsToScreen );
```
I added the `const` keyword here

francoishamon commented on src/coreComponents/physicsSolvers/SolverBase.cpp:
```diff
@@ -75,6 +75,16 @@ SolverBase::SolverBase( string const & name,
+  registerWrapper( viewKeyStruct::computeStatisticsString(), &m_computeStatistics ).
+    setApplyDefaultValue( 0 ). // do nothing by default
+    setInputFlag( InputFlags::OPTIONAL ).
+    setDescription( "Flag indicating whether statistics are computed or not" );
```
In the new version of the PR, I would like to use `TimeHistory` to output the statistics over time (i.e., no output to terminal, no output to vtk). So, I think it would be better not to use `plotLevel` or `logLevel` to control the behavior from the XML file.

francoishamon commented:
@TotoGaz I updated this PR based on your comments, and it is ready for another round of reviews when you have some time. I modified the PR to compute the average, min, and max quantities on a per-region basis, which is what our use case requires. Therefore, you will see that the statistics are now registered by the solver on the `ElementRegions`. Let me know what you think.

TotoGaz commented on src/coreComponents/physicsSolvers/NonlinearSolverParameters.hpp:
```diff
@@ -172,14 +175,23 @@ class NonlinearSolverParameters : public dataRepository::Group
   real64 m_timeStepCutFactor;
 
-  /// number of times that the time-step had to be cut
+  /// Number of times that the time-step had to be cut
   integer m_numdtAttempts;
 
   /// number of times that the configuration had to be changed
   integer m_numConfigurationAttempts;
 
   /// Max number of times that the configuration can be changed
   integer m_maxNumConfigurationAttempts;
+
+  // Simulation statistics
+
+  /// Total number of successful Newton iterations
+  integer m_totalSuccessfulNewtonNumIterations;
+
+  /// Total number of wasted Newton iterations
+  integer m_totalWastedNewtonNumIterations;
+
```
This should (at least) go in its own `class`.
Here we're mixing input/configuration parameters and meta-results.
As a side effect you could wrap a `LvArray< integer >`, one index for each information, so you should be able to use it in `TimeHistory` too (I know it's a workaround 😉 )

francoishamon replied:
The PR no longer deals with the solver parameters, I thought it would be easier to do something separately.

TotoGaz commented on src/coreComponents/physicsSolvers/NonlinearSolverParameters.cpp:
```diff
@@ -111,6 +113,8 @@ NonlinearSolverParameters::NonlinearSolverParameters( string const & name,
     setInputFlag( InputFlags::OPTIONAL ).
     setDescription( "Max number of times that the configuration can be changed" );
 
+  registerWrapper( viewKeysStruct::totalSuccessfulNewtonNumIterationsString, &m_totalSuccessfulNewtonNumIterations );
+  registerWrapper( viewKeysStruct::totalWastedNewtonNumIterationsString, &m_totalWastedNewtonNumIterations );
```
If you use one class instead of multiple integers you'll be able to register it once.

TotoGaz commented on src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseBase.cpp:
```diff
@@ -245,19 +247,32 @@ void CompositionalMultiphaseBase::registerDataOnMesh( Group & meshBodies )
+      {
+        ElementRegionBase & region = elemManager.getRegion( regionNames[i] );
+
+        region.registerWrapper< real64 >( viewKeyStruct::averagePressureString() );
+        region.registerWrapper< real64 >( viewKeyStruct::minimumPressureString() );
+        region.registerWrapper< real64 >( viewKeyStruct::maximumPressureString() );
+        region.registerWrapper< real64 >( viewKeyStruct::totalPoreVolumeString() );
+        region.registerWrapper< real64 >( viewKeyStruct::totalUncompactedPoreVolumeString() );
+        region.registerWrapper< array1d< real64 > >( viewKeyStruct::phasePoreVolumeString() ).
+          setSizedFromParent( 0 ).reference().
+          resizeDimension< 0 >( m_numPhases );
```
Same here, could you try to use a utility `class/struct` instead?

TotoGaz commented on src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseFVM.hpp:
```diff
@@ -133,21 +126,20 @@ class CompositionalMultiphaseFVM : public CompositionalMultiphaseBase
-  void
-  computeCFLNumbers( real64 const & dt, DomainPartition & domain );
-
+  virtual void
+  computeStatistics( real64 const & dt,
+                     DomainPartition & domain ) const override;
```
This member functions conditionally computes the statistics if `m_computeStatistics` is `true` (I think that the `doxygen` makes no mention of it).
I would compute the statistics whatever `m_computeStatistics` because it can be clearer.
And I would let the client code check that `m_computeStatistics` is `true` before calling.
If it happens that we have a lot of `if ( m_computeStatistics ) computeStatistics( dt, domain )`, we can make things evolve.

francoishamon replied:
I changed my approach to control the output of the statistics. Now they are computed as part of `Tasks`, so we don't have to pollute the solvers with statistics calculations anymore. In particular, this `computeStatistics` function is now gone.

TotoGaz commented on src/coreComponents/physicsSolvers/fluidFlow/SinglePhaseBaseKernels.hpp:
```diff
@@ -566,6 +566,58 @@ struct SolutionCheckKernel
+  {
+    RAJA::ReduceMin< parallelDeviceReduce, real64 > subRegionMinPres( LvArray::NumericLimits< real64 >::max );
+    RAJA::ReduceSum< parallelDeviceReduce, real64 > subRegionAvgPresNumerator( 0.0 );
+    RAJA::ReduceMax< parallelDeviceReduce, real64 > subRegionMaxPres( 0.0 );
```
This is meant to be positive or can you have negative values too?
In the latter case, starting value `0.0` needs to be changed.
And in the first case maybe too (do we intend this value to be unchanged?)...

francoishamon replied:
You are right, I changed the min value initialization to `-LvArray::NumericLimits< real64 >::max` because the single-phase solver allows for negative values.

TotoGaz commented on src/coreComponents/physicsSolvers/fluidFlow/SinglePhaseBaseKernels.hpp:
```diff
@@ -566,6 +566,58 @@ struct SolutionCheckKernel
+
+      // To match our "reference", we have to use reference porosity here, not the actual porosity when we compute averages
+      real64 const uncompactedPoreVol = volume[ei] * refPorosity[ei];
+      real64 const dynamicPoreVol = volume[ei] * porosity[ei][0];
```
Why index 0 in `porosity[ei][0]`? Could you document it?

francoishamon replied:
This is because the data living in constitutive models has a dimension for the quadrature point index (always equal to zero in the FV solvers). It shows up everywhere in the solvers, so it would be a little overwhelming to comment on this every time.

TotoGaz commented on src/coreComponents/physicsSolvers/fluidFlow/SinglePhaseBase.cpp:
```diff
@@ -262,9 +279,114 @@ void SinglePhaseBase::initializePostInitialConditionsPreSubGroups()
+    arrayView1d< real64 const > const refPorosity = solid.getReferencePorosity();
+    arrayView2d< real64 const > const porosity = solid.getPorosity();
+
+    real64 subRegionAvgPresNumerator = 0.0;
+    real64 subRegionMinPres = 0.0;
+    real64 subRegionMaxPres = 0.0;
+    real64 subRegionTotalUncompactedPoreVol = 0.0;
+    real64 subRegionTotalPoreVol = 0.0;
```
Do these values really make sense?

francoishamon replied:
Yes, it makes sense to initialize the values like this before computing the min and the max over the sub-region

TotoGaz reviewed (commented):
It's surely possible to come up with better `Statistics` classes managemenent.

francoishamon commented on src/coreComponents/physicsSolvers/solidMechanics/SolidMechanicsStatistics.hpp:
```diff
@@ -0,0 +1,110 @@
+    /// minimum displacement in each direction
+    array1d< real64 > minDisplacement;
+    /// maximum displacement in each direction
+    array1d< real64 > maxDisplacement;
```
I added a comment to say that it means in the x, y, and z directions

francoishamon commented on src/coreComponents/physicsSolvers/fluidFlow/SinglePhaseStatistics.hpp:
```diff
@@ -0,0 +1,120 @@
+ *
+ * Task class allowing for the computation of aggregate statistics in single-phase simulations
+ */
+class SinglePhaseStatistics : public TaskBase
```
I implemented the `FieldStatisticsBase` class that you suggested. It is used to automate the registration of the solver as we discussed.

francoishamon commented on src/coreComponents/physicsSolvers/fluidFlow/FlowSolverBase.hpp:
```diff
@@ -144,16 +149,18 @@ class FlowSolverBase : public SolverBase
   real64 m_fluxEstimate;
 
+  /// flag to decide whether statistics (avg pressure, CFL) are computed or not
+  integer m_computeStatistics;
```
Sorry, that was remnant of the old implementation without `Task`. Now this flag is not needed anymore (I removed it) and I use events to trigger the statistics tasks.

francoishamon commented on src/coreComponents/physicsSolvers/solidMechanics/SolidMechanicsStatistics.hpp:
```diff
@@ -0,0 +1,110 @@
+   * @brief Compute node-based statistics on the reservoir
+   * @param[in] mesh the mesh level object
+   */
+  void computeNodeStatistics( MeshLevel & mesh ) const;
```
I made this function `private`

francoishamon replied:
In fact, I had forgotten why I put this function `public` in the first place: it has a kernel launch, so the code does not compile if the function is `private`.

francoishamon commented on src/coreComponents/physicsSolvers/solidMechanics/SolidMechanicsStatistics.cpp:
```diff
@@ -0,0 +1,166 @@
+                                                                        arrayView1d< string const > const & )
+  {
+    NodeManager & nodeManager = mesh.getNodeManager();
+    nodeManager.registerWrapper< NodeStatistics >( viewKeyStruct::nodeStatisticsString() ).
```
Here, I prefer registering the statistics on the `NodeManager` rather than on the `Task` in case there are multiple mesh bodies in the simulation,

TotoGaz commented on src/coreComponents/physicsSolvers/FieldStatisticsBase.hpp:
```diff
@@ -0,0 +1,99 @@
+  }
+
+  /// Name of the solver
+  string m_solverName;
```
The `m_solverName` is used to retrieve the `m_solver`.
Once done, the we could ditch it (while I think that unfortunately, it's not possible in our current pattern). And if not, `m_solver->getName()` will do the trick in a more consistent way.

Anyhow, `m_solverName` can surely be declared `private`.
And maybe `postProcessInput` as well.

francoishamon replied:
I made the `m_solverName` `private`. The `postProcessInput` function cannot be `private` because it is used in the derived classes.

TotoGaz commented on src/coreComponents/physicsSolvers/FieldStatisticsBase.hpp:
```diff
@@ -0,0 +1,99 @@
+      setDescription( "Name of the " + SOLVER::coupledSolverAttributePrefix() + " solver" );
+  }
+
+  /**
+   * @defgroup Tasks Interface Functions
+   *
+   * This function implements the interface defined by the abstract TaskBase class
+   */
+  /**@{*/
+
+  virtual bool execute( real64 const time_n,
+                        real64 const dt,
+                        integer const cycleNumber,
+                        integer const eventCounter,
+                        real64 const eventProgress,
+                        DomainPartition & domain ) override = 0;
+
+  /**@}*/
```
I'm not sure this really is mandatory, after all... 🤷

TotoGaz commented on src/coreComponents/physicsSolvers/solidMechanics/SolidMechanicsStatistics.hpp:
```diff
@@ -0,0 +1,98 @@
+{
+public:
+
+  using Base = FieldStatisticsBase< SolidMechanicsLagrangianFEM >;
```
Do you need this? And do you need it `public`?

francoishamon replied:
It is needed in the constructor of the class for instance. I made it `private`.

TotoGaz replied:
I was able to spot its usage, but since you use it only once in the ctor, I was thinking it was a bit overkill. Anyway, `private` is better.

TotoGaz commented on src/coreComponents/physicsSolvers/fluidFlow/SinglePhaseStatistics.hpp:
```diff
@@ -0,0 +1,109 @@
+{
+public:
+
+  using Base = FieldStatisticsBase< SinglePhaseBase >;
```
Do you need this? And do you need it public?

francoishamon replied:
I made it `private`.

TotoGaz replied:
same

TotoGaz commented on src/coreComponents/physicsSolvers/fluidFlow/SinglePhaseStatistics.hpp:
```diff
@@ -0,0 +1,109 @@
+namespace geosx
+{
+
+class SinglePhaseBase;
```
Does importing this class result in some link error?

francoishamon replied:
I did not get a link error with this forward declaration.

TotoGaz replied:
Indeed, but what about using a proper `#include`?

TotoGaz reviewed (approved):
LGTM
